nmdc_runtime/api/endpoints/find.py

```diff
@@ -160,42 +160,90 @@ def find_data_objects_for_study(
     biosamples = mdb.biosample_set.find({"associated_studies": study["id"]}, ["id"])
     biosample_ids = [biosample["id"] for biosample in biosamples]
 
+    biosample_data_objects = []
+
     for biosample_id in biosample_ids:
         current_ids = [biosample_id]
         collected_data_objects = []
 
+        # Iterate over records in the `alldocs` collection. Look for
+        # records that have the given biosample_id as value on the
+        # `has_input` key/slot. The retrieved documents might also have a
+        # `has_output` key/slot associated with them. Get the value of the
+        # `has_output` key and check if it's type is `nmdc:DataObject`. If
+        # it's not, repeat the process till it is.
         while current_ids:
             new_current_ids = []
             for current_id in current_ids:
+                # Query to find all documents with current_id as input
                 query = {"has_input": current_id}
-                document = mdb.alldocs.find_one(query)
+                documents = mdb.alldocs.find(query)
 
-                if not document:
+                if not documents:
                     continue
 
-                has_output = document.get("has_output")
-                if not has_output:
-                    continue
-
-                for output_id in has_output:
-                    if get_classname_from_typecode(output_id) == "DataObject":
-                        data_object_doc = mdb.data_object_set.find_one(
-                            {"id": output_id}
-                        )
-                        if data_object_doc:
-                            collected_data_objects.append(strip_oid(data_object_doc))
-                    else:
-                        new_current_ids.append(output_id)
+                # There may be multiple documents which satisfy the
+                # above `has_input` "matching" criteria
+                for document in documents:
+                    # retrieve `has_output` value for each of the documents
+                    has_output = document.get("has_output")
+
+                    # Another way in which DataObjects can be related to Biosamples is through the
+                    # `was_informed_by` key/slot. We need to link records from the `workflow_execution_set`
+                    # collection that are "informed" by the same DataGeneration records that created
+                    # the outputs above. Then we need to get additional DataObject records that are
+                    # created by this linkage.
+
+                    # If no has_output, check the document type
+                    if not has_output:
+                        # Check if the document is of type "NucleotideSequencing"
+                        if "NucleotideSequencing" in document.get("type"):
```

Comment:
I've added a [d642a26](https://github.com/microbiomedata/nmdc-runtime/pull/738/commits/d642a26badb89ea9f9c91fbc75c50d37cdc72230) to use the descendants of the `DataGeneration` class instead of hardcoding `NucleotideSequencing`, `MassSpectrometry`, etc.